### PR TITLE
INT & ANN: Implement intention and fix for struct fields expansion instead of `..`

### PR DESCRIPTION
**Pull request by shevtsiv**
In this pull request I intend to do the following:
1) Create E0023 and E0027 error classes and implement analysis for them.
2) Create fix for those errors.
3) Create intention for struct fields expansion.
4) Create a module with shared code for using it in the fix and intention.

Once these steps are complete I will squash all the commits and #3928 should be closed after merge.

### CODE REVIEW

**Comment by shevtsiv** [comment]
Although this is a draft proposals and ideas are welcome.

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.rust.lang.core.psi.ext.descendantsOfType
+import org.rust.lang.core.psi.ext.fields
+import org.rust.lang.core.resolve.ref.deepResolve
+import org.rust.openapiext.buildAndRunTemplate
+import org.rust.openapiext.createSmartPointer
+
+fun expand(project: Project, editor: Editor, body: PsiElement, startElement: PsiElement) {
+    var anchor = startElement.descendantsOfType<LeafPsiElement>().first { it.text == "{" }
+    while (anchor.nextSibling.text != "}") {
+        anchor.nextSibling.delete()
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: It's a bad idea to delete all fields and fill them again b/c there can be comments or specific formatting. Take a look at e.g. `AddStructFieldsFix` (`calculateMissingFields`/`fillStruct`)

File: src/main/kotlin/org/rust/ide/annotator/fixes/ExpandStructFieldsFix.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    private val declaration: RsFieldsOwner
+) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
+    override fun getText() = "Expand actual struct fields"
+
+    override fun getFamilyName() = text
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, startElement: PsiElement, endElement: PsiElement) {
+        if (startElement is RsPatStruct) {
+            val declarationFieldNames = declaration.fields.mapToSet { it.name }
+            val extraFields = startElement.patFieldList.filter { it.text !in declarationFieldNames }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/fixes/ExpandStructFieldsFix.kt]:
```suggestion
            val extraFields = startElement.patFieldList.filter { kind.fieldName!in declarationFieldNames }
```

If you use `element.text` (anywhere), it's probably a bug. This fails on `S { ref field }` or `S { field: f }`.

File: src/main/kotlin/org/rust/ide/annotator/fixes/ExpandStructFieldsFix.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            val hasTrailingComma = startElement.rbrace.getPrevNonCommentSibling()?.elementType == RsElementTypes.COMMA || startElement.dotdot != null
+            extraFields.forEach {
+                it.nextSibling.delete()
+                it.delete()
+            }
+            if (startElement.patFieldList.size >= declarationFieldNames.size) {
+                startElement.dotdot?.delete()
+                return
+            }
+            val bodyFieldNames = startElement.patFieldList.mapToSet { it.text }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/fixes/ExpandStructFieldsFix.kt]:
```suggestion
            val bodyFieldNames = startElement.patFieldList.mapToSet { it.kind.fieldName }
```

File: src/main/kotlin/org/rust/ide/intentions/ExpandStructFieldsIntention.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        } else {
+            null
+        }
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val structBody = ctx.structBody
+        if (structBody is RsPatStruct) {
+            val declaration = structBody.path.reference.deepResolve() as? RsFieldsOwner ?: return
+            val bodyFieldNames = structBody.patFieldList.map { it.text }.toSet()
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/intentions/ExpandStructFieldsIntention.kt]:
```suggestion
            val bodyFieldNames = structBody.patFieldList.map { it.kind.fieldName }.toSet()
```

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 5]
```diff
@@ -84,11 +84,51 @@ class RsErrorAnnotator : RsAnnotatorBase(), HighlightRangeExtension {
...
+            bodyFields.size + 1
+        } else {
+            bodyFields.size
+        }
+        if (bodyFieldsSize != declaration.fields.size) {
+            RsDiagnostic.InvalidStructPattern(rsPatStruct, declaration).addToHolder(holder)
+            return
+        }
+        val differFromDeclaration = { fieldsPair : Pair<RsFieldDecl, RsPatField> ->
+            fieldsPair.first.name != fieldsPair.second.text
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]:
```suggestion
            fieldsPair.first.name != fieldsPair.second.kind.fieldName
```

File: src/main/kotlin/org/rust/ide/annotator/fixes/ExpandStructFieldsFix.kt [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+class ExpandStructFieldsFix(
+    element: PsiElement,
+    private val declaration: RsFieldsOwner
+) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
+    override fun getText() = "Expand actual struct fields"
+
+    override fun getFamilyName() = text
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, startElement: PsiElement, endElement: PsiElement) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/fixes/ExpandStructFieldsFix.kt]: Let's rename `startElement` to `pat` to make code cleaner

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+package org.rust.ide.utils
+
+import com.intellij.psi.PsiElement
+import com.intellij.psi.impl.source.tree.LeafPsiElement
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+fun expand(factory: RsPsiFactory, rsPat: RsPat, missingFields: List<PsiElement>, hasTrailingComma: Boolean) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: Too generic name. Let's rename to something more specific

File: src/main/kotlin/org/rust/ide/annotator/fixes/ExpandStructFieldsFix.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    private val declarationFieldNames: List<String>
+) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
+    override fun getText() = "Expand actual struct fields"
+
+    override fun getFamilyName() = text
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, pat: PsiElement, endElement: PsiElement) {
+        val factory = RsPsiFactory(project)
+        if (pat is RsPatStruct) {
+            val hasTrailingComma = pat.rbrace.getPrevNonCommentSibling()?.elementType == RsElementTypes.COMMA || pat.dotdot != null
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/fixes/ExpandStructFieldsFix.kt]:
```suggestion
            val hasTrailingComma = pat.rbrace.getPrevNonCommentSibling()?.elementType == RsElementTypes.COMMA
```

Looks like ` || pat.dotdot != null` is redundant

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 9]
```diff
@@ -85,11 +85,48 @@ class RsErrorAnnotator : RsAnnotatorBase(), HighlightRangeExtension {
...
+    private fun checkRsPatTupleStruct(holder: AnnotationHolder, rsPatTupleStruct: RsPatTupleStruct) {
+        val declaration = rsPatTupleStruct.path.reference.deepResolve() as? RsFieldsOwner ?: return
+        val bodyFields = rsPatTupleStruct.childrenOfType<RsPatIdent>()
+        if (bodyFields.size < declaration.fields.size) {
+            if (declaration.elementType == RsElementTypes.ENUM_VARIANT) {
+                RsDiagnostic.MissingFieldsInEnumVariantPattern(rsPatTupleStruct, declaration).addToHolder(holder)
+            } else {
+                RsDiagnostic.MissingFieldsInStructPattern(rsPatTupleStruct, declaration).addToHolder(holder)
+            }
+        } else if (bodyFields.size > declaration.fields.size) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]:
```suggestion
        } else if (bodyFields.size > declaration.fields.size && rsPatTupleStruct.dotdot == null) {
```

Tuple patterns can also contain `..`:
```rust
let Foo(a, .., b) = Foo(...);
```

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                }
+            }
+        }
+        fillRsPatWithFields(factory, rsPat, missingFields.drop(insertedFieldsAmount), hasTrailingComma)
+    } else if (rsPat is RsPatTupleStruct) {
+        fillRsPatWithFields(factory, rsPat, missingFields, hasTrailingComma)
+    }
+}
+
+fun getTupleStructMissingFields(factory: RsPsiFactory, amount: Int): List<RsPatBinding> {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: Let's make it private somehow (see subsequent comments) or move it to `RsPsiFactory` and rename with `create` prefix.

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    val missingFields = ArrayList<RsPatBinding>(amount)
+    var tempNamePlaceholder = 'a'
+    for (i in 0 until amount) {
+        missingFields.add(factory.createPatBinding(tempNamePlaceholder.toString()))
+        tempNamePlaceholder++
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: Nice, but can overflow (if more than 26 fields) :)

File: src/main/kotlin/org/rust/ide/intentions/ExpandStructFieldsIntention.kt [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            var position = 0
+            declaration.fields.forEach { fieldPositions[it.name!!] = position++ }
+            expandStructFields(RsPsiFactory(project), structBody, fieldPositions, missingFields)
+        } else if (structBody is RsPatTupleStruct) {
+            val declaration = structBody.path.reference.deepResolve() as? RsFieldsOwner ?: return
+            val bodyFields = structBody.childrenOfType<RsPatIdent>()
+            structBody.dotdot?.delete()
+            val missingFields = getTupleStructMissingFields(factory, declaration.fields.size - bodyFields.size)
+            fillRsPatWithFields(factory, structBody, missingFields)
+            editor.buildAndRunTemplate(ctx.structBody, ctx.structBody.childrenOfType<RsPatBinding>().map { it.createSmartPointer() })
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/intentions/ExpandStructFieldsIntention.kt]: Let's extract these 2 lines to a separate function called `expandTupleStructFields` and put it to `StrunctFieldsExpander.kt`. This is more convenient with "struct" branch, allows to reduce code duplication and also makes available to make `fillRsPatWithFields` private.

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+import com.intellij.psi.PsiElement
+import com.intellij.psi.impl.source.tree.LeafPsiElement
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.stdext.nextOrNull
+
+fun expandStructFields(
+    factory: RsPsiFactory,
+    rsPat: RsPat,
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]:
```suggestion
    pat: RsPatStruct,
```

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    declarationFieldPositions: HashMap<String, Int>,
+    missingFields: List<PsiElement>,
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]:
```suggestion
    missingFields: List<RsPatField>,
```

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+                    insertedFieldsAmount++
+                    break
+                } else {
+                    anchor = patFieldListIterator.nextOrNull()
+                }
+            }
+        }
+        fillRsPatWithFields(factory, rsPat, missingFields.drop(insertedFieldsAmount), hasTrailingComma)
+    } else if (rsPat is RsPatTupleStruct) {
+        fillRsPatWithFields(factory, rsPat, missingFields, hasTrailingComma)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: dead code

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    val missingFields = ArrayList<RsPatBinding>(amount)
+    var tempNamePlaceholder = 'a'
+    for (i in 0 until amount) {
+        missingFields.add(factory.createPatBinding(tempNamePlaceholder.toString()))
+        tempNamePlaceholder++
+    }
+    return missingFields
+}
+
+fun fillRsPatWithFields(factory: RsPsiFactory, rsPat: RsPat, fields: List<PsiElement>, hasTrailingComma: Boolean = false) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: Let's also make it private

File: src/main/kotlin/org/rust/ide/intentions/ExpandStructFieldsIntention.kt [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+                .map { factory.createPatField(it.name!!) }
+            val fieldPositions = HashMap<String, Int>(missingFields.size)
+            var position = 0
+            declaration.fields.forEach { fieldPositions[it.name!!] = position++ }
+            expandStructFields(RsPsiFactory(project), structBody, fieldPositions, missingFields)
+        } else if (structBody is RsPatTupleStruct) {
+            val declaration = structBody.path.reference.deepResolve() as? RsFieldsOwner ?: return
+            val bodyFields = structBody.childrenOfType<RsPatIdent>()
+            val hasTrailingComma = structBody.rparen.getPrevNonCommentSibling()?.elementType == RsElementTypes.COMMA
+            if (structBody.dotdot?.getNextNonCommentSibling()?.elementType == RsElementTypes.COMMA) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/intentions/ExpandStructFieldsIntention.kt]: Consider extracting `structBody.dotdot?.getNextNonCommentSibling()` to a variable

File: src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.ref.deepResolve
+import org.rust.openapiext.buildAndRunTemplate
+import org.rust.openapiext.createSmartPointer
+
+fun expandStructFields(factory: RsPsiFactory, patStruct: RsPatStruct) {
+    val declaration = patStruct.path.reference.deepResolve() as? RsFieldsOwner ?: return
+    val hasTrailingComma = patStruct.rbrace.getPrevNonCommentSibling()?.elementType == RsElementTypes.COMMA
+    patStruct.dotdot?.delete()
+    val existingFields = patStruct.patFieldList
```
Comment (ice1000) [inline src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: Does this `patFieldList` return `_` patterns as well?

Comment (shevtsiv) [reply src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: > Does this `patFieldList` return `_` patterns as well?

Yes, this `patFieldList` also includes underscores, but the `patTupleStruct.childrenOfType<RsPatIdent>()` that was used below does not do this. That's why `_` were considered wrong fields. 

Thank you for the report!

Comment (ice1000) [reply src/main/kotlin/org/rust/ide/utils/StructFieldsExpander.kt]: > but the patTupleStruct.childrenOfType<RsPatIdent>() that was used below does not do this. 

Oh! I didn't catch that. Thanks for your fix!

**Pull request merged**